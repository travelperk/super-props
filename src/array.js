import { error } from "./utils";
import { randomNumber } from "./utils";

export default function array(
  shape,
  {
    nullable = false,
    allowEmpty = true,
    minLength = 0,
    maxLength = Infinity,
    unique = false
  } = {}
) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value == null && nullable) return;

    if (!Array.isArray(value))
      throw error(
        `Property \`${propName}\` of \`${componentName}\` is not an array.`
      );

    if (value.length === 0 && !allowEmpty)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` can't be an empty array.`
      );

    if (value.length < minLength)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` can't be shorter than ${minLength}.`
      );

    if (value.length > maxLength)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` can't be longer than ${maxLength}.`
      );

    value.forEach((item, index) => {
      shape({ [index]: item }, index, `${componentName}[]`);
    });

    if (unique && value.length !== Array.from(new Set(value)).length)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` has repeated values.`
      );
  }

  type.toString = function() {
    return `array(${shape})`;
  };

  type.make = function() {
    const length = randomNumber({
      min: minLength,
      max: maxLength === Infinity ? 10 : maxLength,
      integer: true
    });

    const generatedArray = [];
    for (let i = 0; i < length; i++) {
      let element;
      do {
        element = shape.make();
      } while (unique && isElementInArray(element, generatedArray));
      generatedArray.push(element);
    }
    return generatedArray;
  };

  return type;
}

function isElementInArray(element, array) {
  return array.some(e => e === element);
}
