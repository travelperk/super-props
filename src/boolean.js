import { error, choose } from "./utils";

export default function boolean({ nullable = false } = {}) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value == null && nullable) return;

    if (typeof value !== "boolean")
      throw error(
        `Property \`${propName}\` of \`${componentName}\` is not a boolean.`
      );
  }

  type.toString = function() {
    return "boolean()";
  };

  type.make = function make() {
    return choose([true, false]);
  };

  return type;
}
