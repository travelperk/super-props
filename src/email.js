import { string } from ".";

export default function email(options) {
  const pattern = /^[a-z0-9!#$%&'*+/=?^_`{|}~.-]+@[a-z0-9-]+(\.[a-z0-9-]+)*$/i;

  const type = string({ ...options, pattern });

  type.toString = function() {
    return "string()";
  };

  return type;
}
