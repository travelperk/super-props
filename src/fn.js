import { error } from "./utils";

export default function fn({ nullable = false } = {}) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value == null && nullable) return;

    if (typeof value !== "function")
      throw error(
        `Property \`${propName}\` of \`${componentName}\` is not a function.`
      );
  }

  type.toString = function() {
    return "fn()";
  };

  return type;
}
