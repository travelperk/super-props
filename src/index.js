export { default as boolean } from "./boolean";
export { default as number } from "./number";
export { default as integer } from "./integer";
export { default as string } from "./string";
export { default as email } from "./email";
export { default as object } from "./object";
export { default as literal } from "./literal";
export { default as oneOf } from "./oneOf";
export { default as fn } from "./fn";
export { default as array } from "./array";
export { default as thrower } from "./thrower";
