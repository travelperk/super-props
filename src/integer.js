import number from "./number";

export default function integer(opts = {}) {
  const type = number({ integer: true, ...opts });

  type.toString = function() {
    return "integer()";
  };

  return type;
}
