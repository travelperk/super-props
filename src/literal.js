import { boolean } from ".";
import { error } from "./utils";

export default function value(literalValue) {
  if (typeof literalValue === "object" && literalValue !== null)
    throw new Error("literal value can't be an object");

  if (typeof literalValue === "function")
    throw new Error("literal value can't be a function");

  function type(props, propName, componentName) {
    const value = props[propName];

    if (value !== literalValue)
      throw error(
        `Property \`${propName}.\` of \`${componentName}\` was supposed to be \`${literalValue}\` but was \`${value}\`.`
      );
  }

  type.make = function() {
    return literalValue;
  };

  type.toString = function() {
    return `literal(${literalValue})`;
  };

  return type;
}
