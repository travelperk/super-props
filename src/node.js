import PropTypes from "prop-types";
import { error } from "./utils";

export default function node({ nullable = false } = {}) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value === null) return;

    const validator = nullable ? PropTypes.node : PropTypes.node.isRequired;

    const validationError = validator(
      props,
      propName,
      componentName,
      "",
      "",
      "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED"
    );

    if (validationError) throw new error(validationError.message);
  }

  type.toString = function() {
    return "node()";
  };

  return type;
}
