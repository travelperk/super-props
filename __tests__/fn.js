import { fn } from "../src";

describe("fn()", () => {
  it.each`
    value        | nullable | shouldThrow
    ${() => {}}  | ${true}  | ${false}
    ${undefined} | ${false} | ${true}
    ${undefined} | ${true}  | ${false}
    ${null}      | ${false} | ${true}
    ${null}      | ${true}  | ${false}
    ${1}         | ${true}  | ${true}
  `(
    "should generate a validator for function values",
    ({ value, nullable, shouldThrow }) => {
      const expectation = expect(() =>
        fn({ nullable })({ value }, "value", "Component")
      );
      shouldThrow
        ? expectation.toThrowErrorMatchingSnapshot()
        : expectation.not.toThrow();
    }
  );
});
