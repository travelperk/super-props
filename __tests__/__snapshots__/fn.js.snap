// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fn() should generate a validator for function values 1`] = `"[Super Props] Property \`value\` of \`Component\` is not a function."`;

exports[`fn() should generate a validator for function values 2`] = `"[Super Props] Property \`value\` of \`Component\` is not a function."`;

exports[`fn() should generate a validator for function values 3`] = `"[Super Props] Property \`value\` of \`Component\` is not a function."`;
